Feature chunk trash bin

wg1k opened the pull request:
Add a safety layer in case of chunk removal by SaunaFS. Instead, the chunks files will go to a temporary trash bin folder.

wg1k committed:
feat(chunkserver): Add trash bin for chunk files

This commit provides additional safety measure for the system,
temporarilly allowing easier recovery for deleted chunks.

Signed-off-by: Baldor <baldor@leil.io>

wg1k committed:
refactor(chunkserver): Create ChunkTrashManager

For splitting responsibility and better tests.

Signed-off-by: Baldor <baldor@leil.io>

antuan96314 committed:
fix: Prevent gcc-13 compiler error

Signed-off-by: Antuan <antuan@leil.io>

lgsilva3087 commented on src/chunkserver/chunkserver-common/chunk_trash_manager_unittest.cc:
```diff
@@ -0,0 +1,79 @@
+	std::filesystem::path filePath = testDir / "test_file.txt";
+	std::ofstream(filePath) << "dummy content";
+
+	std::string deletionTime = manager.getDeletionTimeString();
```
Consider not to use an instance to access static members:
ChunkTrashManager::getDeletionTimeString();

lgsilva3087 commented on src/chunkserver/chunkserver-common/chunk_trash_manager_unittest.cc:
```diff
@@ -0,0 +1,79 @@
+	std::ofstream(filePath) << "dummy content";
+
+	std::string deletionTime = manager.getDeletionTimeString();
+	int result = manager.moveToTrash(filePath, testDir, deletionTime);
```
Same here and in the others.

lgsilva3087 commented on src/chunkserver/chunkserver-common/chunk_trash_manager_unittest.cc:
Please apply automatic code formatting to the file.

lgsilva3087 commented on src/chunkserver/chunkserver-common/cmr_disk.cc:
```diff
@@ -1,16 +1,17 @@
 #include "cmr_disk.h"
 
 #include <sys/statvfs.h>
+#include <c++/13/iostream>
```
Please review this line, we should not include specific versions of the standard library headers.

lgsilva3087 commented on src/chunkserver/chunkserver-common/cmr_disk.h:
```diff
@@ -129,4 +137,7 @@ class CmrDisk : public FDDisk {
 	                   off64_t offset) override;
+
+private:
+	ChunkTrashManager trashManager;
```
All functions in ChunkTrashManager are static, you probably don't need an instance.

lgsilva3087 commented on src/chunkserver/chunkserver-common/cmr_disk.h:
```diff
@@ -1,7 +1,15 @@
+
 #include "common/platform.h"
 
+#include "chunk_trash_manager.h"
```
If you remove the instance, you can move this include to the .cc.

lgsilva3087 commented on src/chunkserver/chunkserver-common/cmr_disk.cc:
```diff
@@ -173,17 +176,36 @@ void CmrDisk::open(IChunk *chunk) {
+	const std::filesystem::path dataFile = chunk->dataFilename();
+
+	// Use the metaPath() and dataPath() to get the disk paths
+	const std::string metaDiskPath = FDDisk::metaPath();
+	const std::string dataDiskPath = FDDisk::dataPath();
```
Consider using direclty metaPath() and dataPath(). Calling it with the FDDisk looks gives the impression of static and it is not needed in this context.

lgsilva3087 commented on src/chunkserver/chunkserver-common/cmr_disk.cc:
```diff
@@ -173,17 +176,36 @@ void CmrDisk::open(IChunk *chunk) {
+
+	// Ensure we found a valid disk path
+	if (metaDiskPath.empty() || dataDiskPath.empty()) {
+		safs_pretty_syslog(LOG_WARNING, "Error finding disk path for chunk: %s", chunk->metaFilename().c_str());
+		return SAUNAFS_ERROR_ENOENT;
```
Consider safs_pretty_errlog and LOG_ERROR if an error is returned.

lgsilva3087 commented on src/chunkserver/chunkserver-common/cmr_disk.h:
```diff
@@ -1,7 +1,15 @@
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Warray-bounds"
+#pragma GCC diagnostic ignored "-Wstringop-overflow"
+#include <filesystem>
```
Not used.

lgsilva3087 commented on src/chunkserver/chunkserver-common/cmr_disk_unittest.cc:
```diff
@@ -0,0 +1,127 @@
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Warray-bounds"
+#pragma GCC diagnostic ignored "-Wstringop-overflow"
+#include <gtest/gtest.h>
+#pragma GCC diagnostic pop
+#include <fstream>
+#include "chunk_trash_manager.h"
+#include "cmr_disk.h"
+#include "errors/saunafs_error_codes.h"
+#include "chunk_with_fd.h"
```
Please group and sort the includes. Usually we:
- Include common/platform.h first.
- Include standard headers in a separate group.
- Include our files prefixed with the module for better sorting:
  chunkserver-common/chunk_trash_manager.h
  chunkserver-common/chunk_with_fd.h
  chunkserver-common/cmr_disk.h
  errors/saunafs_error_codes.h

lgsilva3087 commented on src/chunkserver/chunkserver-common/disk_with_fd.h:
```diff
@@ -194,6 +194,7 @@ class FDDisk : public IDisk {
 
 	std::string metaPath_;  ///< Metadata directory
 	std::string dataPath_;  ///< Data directory
+	static std::string trashDir_;  ///< Trash directory name
```
Not used.

lgsilva3087 commented on tests/ci_build/run-build.sh:
```diff
@@ -100,7 +100,7 @@ cmake -B "${build_dir}" \
 
 if [ -f "${build_dir}/CPackConfig.cmake" ]; then
-	nice cpack -B "${build_dir}" --config "${build_dir}/CPackConfig.cmake" -j "$(nproc)"
+: 	nice cpack -B "${build_dir}" --config "${build_dir}/CPackConfig.cmake" -j "$(nproc)"
```
Is this related to the change?

lgsilva3087 reviewed (changes requested):
Nice change, please check my suggestions.
Also, I see very nice unit tests but no integration tests. In this case, please consider to add it to see how it works in a real scenario.
The test could:
- Create some files.
- Set small trash time.
- Wait for them to be deleted.
- Check that the files were moved to the expected directory.

wg1k committed:
fix: Address PR sugestions

Co-authored-by: Antuan <antuan@leil.io>
Co-authored-by: Guillex <guillex@leil.io>
Signed-off-by: Baldor <baldor@leil.io>

wg1k committed:
test: Update integration test for unlink

Co-authored-by: Antuan <antuan@leil.io>
Signed-off-by: Baldor <baldor@leil.io>

rolysr commented on tests/test_suites/ShortSystemTests/test_unlink.sh:
```diff
@@ -24,3 +24,9 @@ timeout="3 minutes"
+
+# Ensure thr "unlinked" files are trashed
+trashed_chunk_files=$(find_all_trashed_chunks | wc -l)
+if [ "${trashed_chunk_files}" -eq 0 ]; then
```
Please, consider in this test to use probably assert_equals or some similar approach for checking if the number of chunks in trash was the expected one, instead of just checking if there is some trash or not.

rolysr replied:
Yes, I consider it good now. Just take into account for comparing expected values with some variable value or function output to follow the conventions on other tests by performing this comparisons by using `assert_equals` or `assert_not_equal` . This is a minor change as it is most for having same coding style across all integration tests.

rolysr reviewed (commented):
LGTM. Please consider my minor suggestion for the integration test.

dmga44 reviewed (commented):
Very nice change, though I think it can be improved.

From my understanding, the chunks are never really unlinked, just moved to the trash folder. We could set a trashtime for those chunks, to make sure they are not filling the drives forever.

Maybe this could be an idea for a next iteration of the chunk trash bin feature.

ralcolea reviewed (dismissed):
LGTM! :+1: :fire: :rocket: 
I’ve added a few minor suggestions for your consideration, but they’re not blockers for merging.

dmga44 commented:
> The trash time management is a responsibility for the Garbage collector.

What Garbage collector?

wg1k committed:
fix: Address PR suggestions (2nd wave)

Signed-off-by: Baldor <baldor@leil.io>

ralcolea commented on src/chunkserver/chunkserver-common/cmr_disk_unittest.cc:
```diff
@@ -0,0 +1,159 @@
+
+class CmrDiskTest : public ::testing::Test {
+protected:
+	CmrDiskTest()
```
Perfect!!

ralcolea reviewed (dismissed):
Great job @ictus4u :+1: :fire: :rocket:

wg1k committed:
fix(test): Consider replication delay

Signed-off-by: Baldor <baldor@leil.io>

uristdwarf commented on src/chunkserver/chunkserver-common/cmr_disk.cc:
```diff
@@ -173,17 +180,37 @@ void CmrDisk::open(IChunk *chunk) {
+	const std::filesystem::path dataFile = chunk->dataFilename();
+
+	// Use the metaPath() and dataPath() to get the disk paths
+	const std::string metaDiskPath = metaPath();
+	const std::string dataDiskPath = dataPath();
+
+	// Ensure we found a valid disk path
+	if (metaDiskPath.empty() || dataDiskPath.empty()) {
+		safs_pretty_errlog(LOG_ERR, "Error finding disk path for chunk: %s",
+		                   chunk->metaFilename().c_str());
+		return SAUNAFS_ERROR_ENOENT;
+	}
+
+	// Create a deletion timestamp
+	const std::string deletionTime = ChunkTrashManager::getDeletionTimeString();
 
-	if (::unlink(chunk->metaFilename().c_str()) != 0) {
-		result = -1;
+	// Move meta file to trash
+	int result = ChunkTrashManager::moveToTrash(metaFile, metaDiskPath, deletionTime);
+	if (result != SAUNAFS_STATUS_OK) {
+		return result;
 	}
 
-	if (::unlink(chunk->dataFilename().c_str()) != 0) {
-		result = -1;
+	// Move data file to trash
+	result = ChunkTrashManager::moveToTrash(dataFile, dataDiskPath, deletionTime);
+	if (result != SAUNAFS_STATUS_OK) {
+		return result;
 	}
 
-	return result;
+	return SAUNAFS_STATUS_OK;
```
Consider adding a config option regarding the new behavior and old behavior, with the old behaviour being default. This is quite a change in behavior and can cause confusion and unexpected results ("why is chunkserver not deleting chunks?") for existing installations.

uristdwarf replied:
Under normal  circumstances, I don't believe the older deletion operation is inherently unsafe. The two main causes of accidental deletions have been 1) Upgrades to newer versions and 2) Migrations from LizardFS, both of which are unusual (not a day-to-day activity).

The other thing is that in general, new features should be disabled by default, not only because they haven't been field-tested but also because they might interfere with existing functionality, in this case the deletion of chunks. When bumping major versions we can revisit whether something should be enabled by default.

However, the point of it being safer in general is persuasive enough that I'd go forgo the above and have it enabled by default, but only when the feature is on par with the old behavior (it actually eventually deletes the chunks). The main idea being we should not break existing setups that rely on the chunks actually being deleted. 

Once the feature is fully-complete, we could come back and decide to make it on by default. Until then, we need to inform users/packagers that this is an option, that we encourage usage of it (but they need to be aware that it's not deleting chunks automatically, they have to do it manually) and let them decide whether to use it or not.

uristdwarf commented on src/chunkserver/chunkserver-common/chunk_trash_manager.cc:
```diff
@@ -0,0 +1,55 @@
+int ChunkTrashManager::moveToTrash(const std::filesystem::path &filePath,
+                                   const std::filesystem::path &diskPath,
+                                   const std::string &deletionTime) {
+	if (!std::filesystem::exists(filePath)) {
+		return SAUNAFS_ERROR_ENOENT;
+	}
+
+	const std::filesystem::path trashDir = diskPath / kTrashDirname;
+	std::filesystem::create_directories(trashDir);
+
+	if (!filePath.string().starts_with(diskPath.string())) {
+		return SAUNAFS_ERROR_EINVAL;
+	}
+
+	const std::filesystem::path trashPath =
+			trashDir / (filePath.filename().string() + "." + deletionTime);
+
+	try {
+		std::filesystem::rename(filePath, trashPath);
+	} catch (const std::filesystem::filesystem_error &e) {
+		return SAUNAFS_ERROR_IO;
+	}
```
Please add logging here, instead of just passing errors down (which makes it difficult do find where the errors originated), these are rather serious errors.

uristdwarf reviewed (changes requested):
> Currently WIP

In this case this PR should be marked as draft or the feature should be behind a feature flag (see my other comments in the code). I would prefer feature flag because this is already a very useful in high risk operations (upgrading, migrating from LizardFS etc.) and we should avoid long feature branches that take months.

I already love this feature (you could already achieve the deletion with a cron job), but putting a unfinished feature, that changes how chunkserver behaves significantly (it doesn't delete chunks, so chunkserver can get full over time without emptying the trash) and that is on by default on into dev is not something that should NOT be accepted at ALL in my opinion.

It's fine to have unfinished features in dev and even releases, but they should be disabled by default (for example prometheus is woefully unfinished, but needs to enabled specifically) so they don't affect anyone.

lgsilva3087 commented on src/chunkserver/chunkserver-common/chunk_trash_manager_unittest.cc:
```diff
@@ -0,0 +1,96 @@
+// Test getDeletionTimeString() to ensure it produces a UTC timestamp
+TEST_F(ChunkTrashManagerTest, GetDeletionTimeStringTest) {
+	std::string timestamp = ChunkTrashManager::getDeletionTimeString();
+	EXPECT_EQ(timestamp.size(), strlen("YYYYMMDDHHMMSS"));  // Check that the timestamp has the correct length
+	EXPECT_TRUE(std::all_of(timestamp.begin(), timestamp.end(), ::isdigit));  // Check that all characters are digits
```
Please avoid the long lines in the complete file.

lgsilva3087 commented on src/chunkserver/chunkserver-common/cmr_disk.cc:
```diff
@@ -173,17 +180,37 @@ void CmrDisk::open(IChunk *chunk) {
+	// Move meta file to trash
+	int result = ChunkTrashManager::moveToTrash(metaFile, metaDiskPath, deletionTime);
+	if (result != SAUNAFS_STATUS_OK) {
+		return result;
```
Consider logging the error here.

lgsilva3087 commented on src/chunkserver/chunkserver-common/cmr_disk.cc:
```diff
@@ -173,17 +180,37 @@ void CmrDisk::open(IChunk *chunk) {
+	// Move data file to trash
+	result = ChunkTrashManager::moveToTrash(dataFile, dataDiskPath, deletionTime);
+	if (result != SAUNAFS_STATUS_OK) {
+		return result;
```
Consider logging the error here.

lgsilva3087 commented on src/chunkserver/chunkserver-common/cmr_disk_unittest.cc:
```diff
@@ -0,0 +1,159 @@
+
+TEST_F(CmrDiskTest, UnlinkChunkSuccessful) {
+	// Create a dummy meta and data file in the test directory
+	std::filesystem::path const metaFile = std::filesystem::path(cmrDiskInstance.metaPath()) / "chunk_meta_file.txt";
+	std::filesystem::path const dataFile = std::filesystem::path(cmrDiskInstance.dataPath()) / "chunk_data_file.txt";
```
Avoid the long lines in the complete file.

lgsilva3087 commented on tests/test_suites/ShortSystemTests/test_unlink.sh:
Make sure to run the SMR version of this test.

lgsilva3087 reviewed (dismissed):
LGTM. Please see my comments.
Also: Consider Urmas' suggestion to add it as an option.

wg1k committed:
feat: Add chunk's trash garbage collection

Implement Chunk Trash garbage collection with configuration-based
trash management, including methods for moving files to trash,
deleting expired files, and freeing space based on disk thresholds.

Co-authored-by: Antuan <antuan@leil.io>
Signed-off-by: Baldor <baldor@leil.io>

antuan96314 commented on tests/test_suites/ShortSystemTests/test_unlink.sh:
```diff
@@ -17,10 +17,26 @@ saunafs setgoal xor3 "$xorfile"
 # Wait for removing all the chunks
 timeout="3 minutes"
 if ! wait_for '[[ $(find_all_chunks | wc -l) == 0 ]]' "$timeout"; then
 	test_add_failure $'The following chunks were not removed:\n'"$(find_all_chunks)"
```
Consider a more descriptive message as removal could take more than 3 minutes in some cases. For instance consider setting this master param to low value to trigger much faster chunk operations: `OPERATIONS_DELAY_INIT = 2`

wg1k replied:
tx

antuan96314 reviewed (dismissed):
LGTM

wg1k committed:
test(ChunkTrash): Apply sugestions from PR

Signed-off-by: Baldor <baldor@leil.io>

rolysr commented on tests/test_suites/ShortSystemTests/test_unlink_trash.sh:
```diff
@@ -0,0 +1,71 @@
+# Ensure the "unlinked" files are trashed
+trashed_chunks_count=$(find_all_trashed_chunks | wc -l)
+echo "Trashed chunks count: ${trashed_chunks_count}"
+if [ "${trashed_chunks_count}" -eq 0 ]; then
+	test_add_failure $'The removed chunks were not moved to the trash folder'
+fi
```
Consider this kind of comparisons to be treated like this on the tests
```
# Ensure the "unlinked" files are trashed
assert_equals "0" "${trashed_chunks_count}"
```
In order to be conventionally align to the other tests.

wg1k replied:
ok

wg1k committed:
fix: Solve rebase conflicts

Signed-off-by: Baldor <baldor@leil.io>

wg1k committed:
fix(tests): Address broken tests

- ShortSystemTests.test_unlink_trash
    The addition of the CHUNK_TRASH_FREE_SPACE_THRESHOLD_GB as higher
    priority for the GC was making it to ignore the expiration time
    with the default value for the reduced size test drive.
- ShortSystemTests.test_chunkserver_start_with_damaged_disk
    The attempt to create the trash folder on a simulated damaged disk
    was making this test to fail.
- ChunkTrashIndex.RemoveFileEntry2 (removed)
    It was testing a dismised functionality od removing empty folders.
- saunafs.sh
    Wrong regex did not select the trash folder
- GTest update for using vcpkg. Addressed type casting warnings.
- Typos and format

Signed-off-by: Baldor <baldor@leil.io>

wg1k commented on tests/test_suites/ShortSystemTests/test_unlink.sh:
```diff
@@ -17,10 +17,26 @@ saunafs setgoal xor3 "$xorfile"
+fi
+# Below value is non-deterministic, but it should be enough to test the
+# feature. Avoiding to use a fixed value to prevent false positives.
+random_files_count_before_removal=$(find_all_chunks | wc -l)
```
outdated

lgsilva3087 commented on src/admin/dump_config_command.cc:
```diff
@@ -171,6 +171,11 @@ const static std::unordered_map<std::string, std::string> defaultOptionsCS = {
+    {"CHUNK_TRASH_EXPIRATION_SECONDS", "259200"},
+    {"CHUNK_TRASH_FREE_SPACE_THRESHOLD_GB", "1024"},
+    {"CHUNK_TRASH_GC_BATCH_SIZE", "1000"},
+    {"CHUNK_TRASH_GC_SPACE_RECOVERY_BATCH_SIZE", "100"},
```
This one is documented in data/sfschunkserver.cfg.in but I don't see it in doc/sfschunkserver.cfg.5.adoc. Please verify.

wg1k replied:
done

lgsilva3087 commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,300 @@
+size_t ChunkTrashManagerImpl::kAvailableThresholdGB = 10;
+size_t ChunkTrashManagerImpl::kTrashTimeLimitSeconds = 259200;
+size_t ChunkTrashManagerImpl::kTrashGarbageCollectorBulkSize = 1000;
+size_t ChunkTrashManagerImpl::kGarbageCollectorSpaceRecoveryStep = 10;
```
We usually use 'k' for constants, I see you change them later from config. I suggest to use them like:
size_t kDefaultAvailable... = 10
and then
size_t availableTresh... = your configured value

This way, the initial constant value, which is used in the configuration dumping, is always constant.

wg1k replied:
done

lgsilva3087 commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,300 @@
+size_t ChunkTrashManagerImpl::kAvailableThresholdGB = 10;
+size_t ChunkTrashManagerImpl::kTrashTimeLimitSeconds = 259200;
+size_t ChunkTrashManagerImpl::kTrashGarbageCollectorBulkSize = 1000;
+size_t ChunkTrashManagerImpl::kGarbageCollectorSpaceRecoveryStep = 10;
```
I think I saw 100 for this value in the documentation. Please double check.

wg1k replied:
done

lgsilva3087 commented on doc/sfschunkserver.cfg.5.adoc:
```diff
@@ -153,6 +153,22 @@ the operation is considered failed and is immediately aborted (default: 1000)
+*CHUNK_TRASH_EXPIRATION_SECONDS*:: specifies the timeout in seconds for chunks to remain
+in the trash before being permanently deleted. (Default: 259200)
+
+*CHUNK_TRASH_FREE_SPACE_THRESHOLD_GB*:: sets the available space threshold in
```
1 TiB could be big for NVMe drives.

wg1k replied:
Probably we'll end with different configs for NVME and HDD, but this can be addressed in another task.

Let's set the default value to 0, and expand the docs with a suggestion to set a value like the 10% of their smaller HDD/NVME in the config. Would you agree?

wg1k replied:
done

lgsilva3087 commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,300 @@
+
+namespace fs = std::filesystem;
+
+size_t ChunkTrashManagerImpl::kAvailableThresholdGB = 10;
```
I think this one was 1 TiB in the dumpConfig command and in the documentation. Please double check.

wg1k replied:
done

lgsilva3087 commented on src/data/sfschunkserver.cfg.in:
```diff
@@ -184,6 +184,29 @@
+## Sets the available space threshold in gigabytes. If the available space on
+## the disk falls below this threshold, the system will start deleting older
+## chunks from the trash to free up space. (Default: 1024)
+# CHUNK_TRASH_FREE_SPACE_THRESHOLD_GB = 1024
```
Please check the other comment about 1 TiB here.

wg1k replied:
Answered there

lgsilva3087 commented on tests/test_suites/ShortSystemTests/test_unlink.sh:
```diff
@@ -17,6 +18,7 @@ saunafs setgoal xor3 "$xorfile"
 dd if=/dev/zero of="$file" bs=1MiB count=130
 dd if=/dev/zero of="$xorfile" bs=1MiB count=130
 saunafs settrashtime 0 "$file" "$xorfile"
+chunks_count_before_files_removal="$(find_all_chunks | wc -l)"
```
Is this value used?

wg1k replied:
~yes~

~`assert_success test "${trashed_chunks_count}" -ge "${chunks_count_before_files_removal}"`~

Edit: No, you are right. I've moved the test to a new file and forgot that line. :+1: 
I am removing it.

lgsilva3087 commented on tests/test_suites/ShortSystemTests/test_unlink_trash.sh:
```diff
@@ -0,0 +1,56 @@
+			`|OPERATIONS_DELAY_INIT = 1" \
+	CHUNKSERVER_EXTRA_CONFIG="CHUNK_TRASH_ENABLED=1`
+			`|CHUNK_TRASH_EXPIRATION_SECONDS=50`
+			`|CHUNK_TRASH_FREE_SPACE_THRESHOLD_GB=0" \
```
This option could be tested, maybe in other test, making sure that the space is immediately recovered if needed.

wg1k replied:
Yes, we can automate another test, but it was tested. The space was immediately recovered, that was exactly making this test to fail, cause the waiting was not working because of the big threshold compared to our test disk size :).

wg1k replied:
Added `ShortSystemTests.test_unlink_trash_no_available_space`

lgsilva3087 reviewed (commented):
Initial review. Please see my comments.

wg1k committed:
feat: Change CHUNK_TRASH_ENABLED default

By default the feature will be disabled.

Signed-off-by: Baldor <baldor@leil.io>

wg1k committed:
fix: Apply suggestions from peer review

Co-authored-by: Crash <crash@leil.io>
Co-authored-by: guillex <guillex@leil.io>
Co-authored-by: Roly <rolysr@leil.io>
Co-authored-by: Urmas Rist <urmas@@leil.io>
Signed-off-by: Baldor <baldor@leil.io>

uristdwarf commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,297 @@
+
+int ChunkTrashManagerImpl::removeFileFromTrash(const std::string &filePath) {
+	if (!isTrashPath(filePath)) {
+		safs_pretty_syslog(LOG_ERR, "Invalid trash path: %s", filePath.c_str());
```
Should use safs::log_err instead of the C function

wg1k replied:
done

uristdwarf commented on doc/sfschunkserver.cfg.5.adoc:
```diff
@@ -153,6 +153,27 @@ the operation is considered failed and is immediately aborted (default: 1000)
+suggested value is about the 15% of the capacity of the smaller configured disk.
+(Default: 0)
+
+*CHUNK_TRASH_GC_BATCH_SIZE*:: defines the bulk size for the garbage collector
+when processing chunks in the trash. This determines how many files are
+processed in each garbage collection cycle. (Default: 1000)
+
+*CHUNK_TRASH_GC_SPACE_RECOVERY_BATCH_SIZE*:: [ADVANCED] The number of files to
+remove from the trash in a single GC cycle, in case the disk is full and space
+needs to be recovered. (Default: 100)
```
Why is RECOVERY_BATCH_SIZE lower than the normal BATCH_SIZE? To me, it would make sense if we are running out of space, we need to delete much more than in our regular GC cycle.

wg1k replied:
I understand why this might seem counterintuitive. The key difference lies in the context of each setting:

CHUNK_TRASH_GC_BATCH_SIZE is part of the regular garbage collection process, handling expired chunk files. Since this process is expected to remove a moderate number of files at a time, a larger batch size ensures efficient cleanup without excessive overhead.

CHUNK_TRASH_GC_SPACE_RECOVERY_BATCH_SIZE is triggered when disk space runs low and operates continuously until CHUNK_TRASH_FREE_SPACE_THRESHOLD_GB is met. While it might seem logical to delete as much as possible in this scenario, a smaller batch size helps prevent excessive I/O and CPU spikes, which could degrade system performance. Additionally, if trash was previously disabled and then re-enabled, the system might need to remove a massive amount of data. A controlled, smaller batch size ensures sustained cleanup without overwhelming the system, allowing it to free up space while maintaining stability.

uristdwarf commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,297 @@
+	std::error_code errorCode;
+	fs::remove(filePath, errorCode);  // Remove the file or directory
+	if (errorCode) {
+		safs_pretty_syslog(LOG_ERR, "Failed to remove file or directory: %s",
```
Should use safs::log_err instead of the C function

wg1k replied:
done

uristdwarf commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,297 @@
+
+	// After processing subdirectories, check if the directory is now empty
+	if (fs::is_empty(directory)) { removeFileFromTrash(directory); }
+}
```
There may be permission and other FS exceptions from these functions

wg1k replied:
Tx for pointing to this method. It was dead code (`cleanEmptyFolders` is not used anymore). Removed.

uristdwarf commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,297 @@
+
+	if (!isTrashPath(directory)) {
+		return;  // We only clean up trash directories
+	}
```
Why would it be anything other than isTrashPath? Seems like an assert would fit here

wg1k replied:
It is critical not to allow this component (chunk trash manager) to remove anything outside their folder. Regarding assert, it depends on someone not aware using `CMAKE_BUILD_TYPE=Release`, or removing the  
```
string(REPLACE "-DNDEBUG" "" CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO}")
```
line from the config (which is far away from this code), for the assert not to be available on release anymore. I'd rather not to take that risk. (assert is not meant for release, btw).

uristdwarf replied:
Could you not add both assert and return?

uristdwarf replied:
Or maybe throw an exception

uristdwarf replied:
Regardless, if a programmer calls this function, then a assert is better because when the programmer is testing, it will be likely he will find an assert. And calling this function is a programmer error, not a user error.

Also, I do know we have asserts that work in production. I was recently testing migration and chunkserver asserted on a wrong chunk filename and crash (and crash is the correct course here, as there is a terrible mistake in the program).

uristdwarf replied:
At the very least, it should not be a silent error, but sirens blaring because something is terribly wrong.

wg1k replied:
> ...we have asserts that work in production

Because of the `-DNDEBUG` line I mentioned above.

> it should not be a silent error

I agree, tx. It is no silent anymore, a log is generated now and an error code is returned.

uristdwarf commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,297 @@
+size_t ChunkTrashManagerImpl::checkAvailableSpace(const std::string &diskPath) {
+	struct statvfs stat {};
+	if (statvfs(diskPath.c_str(), &stat) != 0) {
+		safs_pretty_syslog(LOG_ERR, "Failed to get file system statistics");
```
Should use safs::log_err instead of the C function

wg1k replied:
done

uristdwarf commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,297 @@
+			}
+			trashIndex->add(deletionTime, file.path().string(), diskPath);
+		}
+	}
```
Should use safs::log_err instead of the C function

wg1k replied:
done

uristdwarf commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,297 @@
+		}
+	}
+
+	trashIndex->reset(diskPath);
```
Should use safs::log_err instead of the C function

Also, we are not printing the error code

wg1k replied:
done, with error code

uristdwarf commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,297 @@
+	if (filePath.find(sourceRoot) != 0) {
+		safs_pretty_syslog(LOG_ERR,
+		                   "File path does not contain source root: %s",
+		                   filePath.c_str());
```
Should use safs::log_err instead of the C function

wg1k replied:
done

uristdwarf commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,297 @@
+	}
+
+	const fs::path trashDir = getTrashDir(diskPath);
+	fs::create_directories(trashDir);
```
Exception may probably occur

wg1k replied:
Changed to the overload with error code.

uristdwarf commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,297 @@
+
+	std::ostringstream oss;
+	oss << std::put_time(utcTime, kTimeStampFormat.c_str());
+	return oss.str();
```
Although very unlikely, gmtime may return NULL and set errno to EOVERFLOW. It is also not thread-safe (according to glibc documentation `man 3 gmtime`)

wg1k replied:
changed to gmtime_r

uristdwarf reviewed (commented):
Review part 1/?

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_index.cc:
```diff
@@ -0,0 +1,127 @@
+ */
+
+#include "chunk_trash_index.h"
+#include "chunk_trash_manager.h"
```
It looks like `chunk_trash_manager.h` is not used and can be safely removed.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_index.cc:
```diff
@@ -0,0 +1,127 @@
+void ChunkTrashIndex::removeInternal(const time_t &deletionTime,
+                                     const std::string &filePath,
+                                     const std::string &diskPath) {
+	auto range = trashIndex[diskPath].equal_range(deletionTime);
```
Consider using [lower_bound](https://cplusplus.com/multimap::lower_bound) instead of `equal_range`.
Actually, the first iterator of `equal_range` is lower_bound.

wg1k replied:
`lower_bound(deletionTime)` is *not always enough* because it only finds the first element >= `deletionTime`, but it does not guarantee the end of the range for that key. Since `trashIndex[diskPath]` is a `std::multimap<std::time_t, std::string>`, *multiple elements can have the same* `deletionTime`. To correctly iterate over all matching entries, we need *both the start and the end of that range*.

ralcolea replied:
Nice catch. I did not see this case you mentioned. Good job :+1: :muscle:

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_index.cc:
```diff
@@ -0,0 +1,127 @@
+
+size_t ChunkTrashIndex::getExpiredFilesInternal(
+    const std::filesystem::path &diskPath, const std::time_t &timeLimit,
+    std::unordered_map<std::string, std::multimap<std::time_t, std::string>>
```
Consider using `TrashIndexDiskEntries` instead of `std::unordered_map<> ...` in the signature of the function.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_index.cc:
```diff
@@ -0,0 +1,127 @@
+	auto &diskTrashIndex = trashIndex[diskPath];
+	TrashIndexFileEntries olderFiles;
+	size_t count = 0;
+	for (auto it = diskTrashIndex.begin(); it != diskTrashIndex.end(); ++it) {
```
Consider using a range-based for loop instead.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_index_unittest.cc:
```diff
@@ -0,0 +1,76 @@
+	          "/.trash.bin/path/to/file");
+}
+
+TEST_F(ChunkTrashIndexTest, RemoveFileEntry) {
```
Consider adding a new unit test to validate the function `remove(const time_t &deletionTime, const std::string &filePath)`, i.e., without provinding the disk path parameter.

Below, there is an example:
```cpp
TEST_F(ChunkTrashIndexTest, RemoveFileEntryFromLastDisk) {
	auto &index = ChunkTrashIndex::instance();
	index.add(1234567890, "/path/to/file", "/testDisk");
	index.add(123456, "/path/to/file", "/testDisk2");

        // Expect removing from /testDisk
	index.remove(1234567890, "/path/to/file");

	auto expiredFiles = index.getExpiredFiles(1234567891);
        // Expect /testDisk should be empty
	EXPECT_EQ(expiredFiles["/testDisk"].size(), static_cast<size_t>(0));
        // Expect /testDisk2 has one file
	EXPECT_EQ(expiredFiles["/testDisk2"].size(), static_cast<size_t>(1));
}
```

wg1k replied:
added!

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_unittest.cc:
```diff
@@ -0,0 +1,128 @@
+protected:
+	void SetUp() override {
+		mockImpl = std::make_shared<MockChunkTrashManagerImpl>();
+		ChunkTrashManager::instance(mockImpl).isEnabled = 1;
```
Accessing static members is allowed without calling `instance()` function.
However, in the current implementation it's not possible because injecting custom implementation relies on the `instance()` function.

Consider to move the injection of a custom implementation to another function, like `initialize()` or `setChunkTrashManagerImpl()` to avoid assigning double responsibility to the instance function.

wg1k replied:
done, added the setImpl() method and removed the instance() one.

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_unittest.cc:
```diff
@@ -0,0 +1,128 @@
+	    .WillOnce(testing::Return(0));  // Assuming 0 is a success return
+	// value.
+
+	int result = ChunkTrashManager::instance().moveToTrash(filePath, diskPath,
```
Accessing static members is allowed without calling `instance()` function.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_unittest.cc:
```diff
@@ -0,0 +1,128 @@
+
+	EXPECT_CALL(*mockImpl, init(diskPath)).Times(1);
+
+	ChunkTrashManager::instance().init(diskPath);  // Call the method.
```
Accessing static members is allowed without calling `instance()` function.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_unittest.cc:
```diff
@@ -0,0 +1,128 @@
+TEST_F(ChunkTrashManagerTest, CollectGarbageForwardsCall) {
+	EXPECT_CALL(*mockImpl, collectGarbage()).Times(1);
+
+	ChunkTrashManager::instance().collectGarbage();  // Call the method.
```
Accessing static members is allowed without calling `instance()` function.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_unittest.cc:
```diff
@@ -0,0 +1,128 @@
+TEST_F(ChunkTrashManagerTest, ReloadConfigForwardsCall) {
+	EXPECT_CALL(*mockImpl, reloadConfig()).Times(1);
+
+	ChunkTrashManager::instance().reloadConfig();  // Call the method.
```
Accessing static members is allowed without calling `instance()` function.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_unittest.cc:
```diff
@@ -0,0 +1,128 @@
+	std::filesystem::path filePath = "example.txt";
+	std::filesystem::path diskPath = "/disk/";
+	std::time_t deletionTime = 1234567890;
+	ChunkTrashManager::instance().isEnabled = 0;
```
Accessing static members is allowed without calling `instance()` function.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_unittest.cc:
```diff
@@ -0,0 +1,128 @@
+	EXPECT_CALL(*mockImpl, moveToTrash(filePath, diskPath, deletionTime))
+	    .Times(0);
+
+	int result = ChunkTrashManager::instance().moveToTrash(filePath, diskPath,
```
Accessing static members is allowed without calling `instance()` function.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_unittest.cc:
```diff
@@ -0,0 +1,128 @@
+
+	EXPECT_CALL(*mockImpl, init(diskPath)).Times(1);
+
+	ChunkTrashManager::instance().isEnabled = 0;
```
Accessing static members is allowed without calling `instance()` function.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_unittest.cc:
```diff
@@ -0,0 +1,128 @@
+TEST_F(ChunkTrashManagerTest, DisabledCollectGarbageDoesNotForwardsCall) {
+	EXPECT_CALL(*mockImpl, collectGarbage()).Times(0);
+
+	ChunkTrashManager::instance().isEnabled = 0;
```
Accessing static members is allowed without calling `instance()` function.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_unittest.cc:
```diff
@@ -0,0 +1,128 @@
+	EXPECT_CALL(*mockImpl, reloadConfig()).Times(1);
+
+	ChunkTrashManager::instance().isEnabled = 0;
+	ChunkTrashManager::instance().reloadConfig();  // Call the method.
```
Accessing static members is allowed without calling `instance()` function.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager.h:
```diff
@@ -0,0 +1,112 @@
+	 *
+	 * @return Reference to the singleton instance of ChunkTrashManager.
+	 */
+	static ChunkTrashManager &instance(ImplementationPtr newImpl = nullptr);
```
`ChunkTrashManager` has all its functions and attributes defined as static. I wonder whether makes sense to have this class as a Singleton.

wg1k replied:
It started out as a singleton by design, but over time it evolved into a full-fledged static class. I agree that the singleton is no longer necessary. Thanks for pointing that out.

wg1k replied:
I removed the singleton in that class.

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,321 @@
+
+#include "errors/saunafs_error_codes.h"
+#include "fmt/format.h"
+#include "fmt/ostream.h"
```
It looks like header `"fmt/ostream.h"` is not used and can be safely removed.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc:
```diff
@@ -0,0 +1,338 @@
+    const std::string &filePath, const std::string &sourceRoot,
+    const std::string &destinationRoot, std::string &destinationPath) {
+	auto error_code = SAUNAFS_STATUS_OK;
+	if (filePath.find(sourceRoot) != 0) {
```
The validation I see here is that `sourceRoot` is a prefix of `filePath`.
However, if `filePath` would contain `sourceRoot` as a substring in a position different than 0, the message will still say `File path does not contain source root`, which might not be true.
I suggest you to update log message to match the validation.

wg1k replied:
Message reworded to "File path is outside the source root"

rolysr commented on src/chunkserver/chunkserver-common/chunk_trash_index.h:
```diff
@@ -0,0 +1,193 @@
+	void add(const std::time_t &deletionTime, const std::string &filePath,
+	         const std::string &diskPath);
+
+	/**
+	 * @brief Removes a file entry from the trash index by its deletion time and
+	 * path.
+	 *
+	 * @param deletionTime The time when the file was deleted.
+	 * @param filePath The path of the file being removed.
+	 */
+	void remove(const time_t &deletionTime, const std::string &filePath);
+
+	/**
+	 * @brief Removes a file entry from the trash index for a specific disk
+	 * path.
+	 *
+	 * @param deletionTime The time when the file was deleted.
+	 * @param filePath The path of the file being removed.
+	 * @param diskPath The path of the disk associated with the file.
+	 */
+	void remove(const time_t &deletionTime, const std::string &filePath,
+	            const std::string &diskPath);
```
For this case, consider using distinct function names rather than overloading because:

- The functions serve different purposes (removing from general trash index vs. removing from a specific disk's trash)
- The intent of each function would be clearer with specific names
- It makes the code more maintainable and self-documenting

wg1k replied:
These functions only remove from the trash index, just that this container include also a `diskPath` section as part of the index. Actually both methods has the same purpose (and call a same internal method), the difference is: if that path is provided, it acts as a filter and the method can avoid to search for it. Te overload seems to behave as intended here.

However, by checking this, I found a forgotten `return` from a copy & paste that had to be adapted. Tx.

rolysr commented on src/chunkserver/chunkserver-common/chunk_trash_index_unittest.cc:
```diff
@@ -0,0 +1,76 @@
+TEST_F(ChunkTrashIndexTest, RemoveFileEntry) {
+	auto &index = ChunkTrashIndex::instance();
+	index.add(1234567890, "/path/to/file", "/testDisk");
+	index.remove(1234567890, "/path/to/file", "/testDisk");
```
Consider also to add testing case for the `remove(const time_t &deletionTime, const std::string &filePath)` function. Which is the one that removes a file entry from the trash index by only receiving its deletion time and path.

wg1k replied:
Added: `RemoveFileEntryWithoutDiskPath`

rolysr commented on src/chunkserver/init.h:
```diff
@@ -36,7 +36,8 @@ inline const std::vector<RunTab> earlyRunTabs = {};
     RunTab{initDiskManager, "disk manager"},  // Always before "plugin manager"
-    RunTab{loadPlugins, "plugin manager"}, RunTab{hddInit, "hdd space manager"},
+    RunTab{loadPlugins, "plugin manager"},
+	RunTab{hddInit, "hdd space manager"},
```
Check for tab space difference here.

wg1k replied:
reverting it to previous status (auto formatted)

rolysr commented on tests/test_suites/ShortSystemTests/test_unlink.sh:
```diff
@@ -5,6 +5,7 @@ CHUNKSERVERS=4 \
 	MASTER_EXTRA_CONFIG="CHUNKS_LOOP_MIN_TIME = 1`
 			`|CHUNKS_LOOP_MAX_CPU = 90`
 			`|OPERATIONS_DELAY_INIT = 0" \
+	CHUNKSERVER_EXTRA_CONFIG="CHUNK_TRASH_ENABLED=0" \
```
Consider not specifying this option here as its default value is expected to be 0.

wg1k replied:
...until the default changes (what is expected once the feature reaches maturity). This test is actually for the case with the chunk trash manager disabled. Since this new config parameter appeared, it should stay for warranting the expected behavior on this test.

rolysr replied:
Good point

rolysr commented on src/chunkserver/chunkserver-common/cmr_disk.cc:
```diff
@@ -188,17 +197,61 @@ void CmrDisk::open(IChunk *chunk) {
+		if (::unlink(metaFile.c_str()) != 0) {
+			safs_pretty_errlog(LOG_ERR, "Error unlinking meta file: %s",
+			                   metaFile.c_str());
+			return SAUNAFS_ERROR_UNKNOWN;
```
Just a doubt on this. Is it possible to know here an specific error cause beforehand, so that we can better name the returned error?

wg1k replied:
tx, changed to return SAUNAFS_ERROR_NOTDONE, which is a bit more descriptive, and include errno in the log.

rolysr reviewed (commented):
LGTM. Just some minor suggestions added.

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl_unittest.cc:
```diff
@@ -0,0 +1,282 @@
+	fs::path filePath = testDir / "valid_file.txt";
+	std::ofstream(filePath.string());  // Create a valid file
+	std::time_t deletionTime = 1729259531;
+	std::string const deletionTimeString = "20241018135211";
```
Why not using `getTimeString(deletionTime)` to generate `deletionTimeString`?

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl_unittest.cc:
```diff
@@ -0,0 +1,282 @@
+
+	int const result =
+	    chunkTrashManagerImpl.moveToTrash(filePath, testDir, deletionTime);
+	ASSERT_NE(result, SAUNAFS_STATUS_OK);
```
You should validate you receive `SAUNAFS_ERROR_ENOENT`, which is the expected error in case the removed file does not exist.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl_unittest.cc:
```diff
@@ -0,0 +1,282 @@
+	std::ofstream(filePath.string());
+
+	std::time_t deletionTime = 1729259531;
+	std::string const deletionTimeString = "20241018135211";
```
Use `getTimeString(deletionTime)` function to build `deletionTimeString`.

wg1k replied:
done, while not required as the test case was clear :)

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl_unittest.cc:
```diff
@@ -0,0 +1,282 @@
+
+	int const result =
+	    chunkTrashManagerImpl.moveToTrash(filePath, testDir, deletionTime);
+	ASSERT_NE(result, SAUNAFS_STATUS_OK);  // Expect failure
```
I suggest you to compare `result` with the expected error for this scenario, i.e., `SAUNAFS_ERROR_NOTDONE`.

It looks like something is missing in the logging for this test. Look at the output below:
![image](https://github.com/user-attachments/assets/a60b3467-e67b-4329-8d48-346bcaa28e36)

wg1k replied:
The first one was done, the second one is related to more logs that we are showing now. This specific one will be investigated in another task.

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl_unittest.cc:
```diff
@@ -0,0 +1,282 @@
+
+	int const result =
+	    chunkTrashManagerImpl.moveToTrash(filePath, testDir, deletionTime);
+	ASSERT_NE(result, SAUNAFS_STATUS_OK);  // Expect failure
```
I suggest you to compare `result` with the expected error, i.e., `SAUNAFS_ERROR_ENOENT`.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl_unittest.cc:
```diff
@@ -0,0 +1,282 @@
+			fs::path filePath =
+			    testDir / ("concurrent_file_" + std::to_string(i) + ".txt");
+			std::ofstream(filePath.string());
+			std::time_t deletionTime = 1729259531;
```
I suggest you to define `deletionTime` as a global variable and use `getTimeString(deletionTime)` to build `deletionTimeString` used at the end of the test.

wg1k replied:
tx

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl_unittest.cc:
```diff
@@ -0,0 +1,282 @@
+}
+
+// Testing moving a file to trash
+TEST_F(ChunkTrashManagerImplTest, MoveToTrash) {
```
I don't see relevant differences between this test and `MoveToTrashValidFile`.

wg1k replied:
agreed, but doesn't harm so far ;)

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl_unittest.cc:
```diff
@@ -0,0 +1,282 @@
+	chunkTrashManagerImpl.removeExpiredFiles(std::time(nullptr) -
+	                                         3600);  // 1 hour ago
+
+	ASSERT_FALSE(
```
I think `removeExpiredFiles()` function does not affect the result of the test. As soon as `moveToTrash()` function is called, `expired_file.txt` will be moved to the trash directory, and `fs::exists(expiredFilePath)` will return `false`, as the original file is no longer in its original path.

wg1k replied:
~If The garbage collector is working, it might affect the result of the test.~
It seems this test was incomplete. Fixed.

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl_unittest.cc:
```diff
@@ -0,0 +1,282 @@
+
+	// Simulate low space condition and check removal
+	chunkTrashManagerImpl.makeSpace(1, 1);       // 1 GB threshold
+	ASSERT_FALSE(fs::exists(lowSpaceFilePath));  // Ensure file is removed
```
This test validates that the file is no longer in its original position, which occurs inmediately after calling `moveToTrash()` function.
I suggest you to validate whether `file1.txt` was removed from the `trash` folder and `trashIndex` data structure.

wg1k replied:
Per your request, I've validated the file removal which is in line with the expected outcome of the function. However, `trashIndex` won't be tested here (unit vs integration) for avoiding complexity of the test and relying on implementation details/dependencies.

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl_unittest.cc:
```diff
@@ -0,0 +1,282 @@
+	chunkTrashManagerImpl.collectGarbage();
+
+	ASSERT_FALSE(
+	    fs::exists(garbageFilePath));  // Ensure garbage file is collected
```
I don't understand the validation of `collectGarbage()` function.
The test validates that `garbage_file.txt` is not available in the original path, but this occurs due to `moveToTrash()`, not because of `collectGarbage()`.

wg1k replied:
This  test was buggy and not well designed, as this corresponded more to an integration test, which is already in place. Removing this unit test.

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl_unittest.cc:
```diff
@@ -0,0 +1,282 @@
+TEST_F(ChunkTrashManagerImplTest, CheckAvailableSpace) {
+	size_t availableSpace =
+	    chunkTrashManagerImpl.checkAvailableSpace(testDir.string());
+	ASSERT_GT(availableSpace,
```
I would prefer to compare the value of `checkAvailableSpace()` with the available space returned by other tool like `df` or other similar to evaluate the accuracy of `checkAvailableSpace()` function.

wg1k replied:
I mocked the `statvfs` function for a more accurate test. Tx.

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl_unittest.cc:
```diff
@@ -0,0 +1,282 @@
+}
+
+// Testing makeSpace with specific disk path
+TEST_F(ChunkTrashManagerImplTest, MakeSpaceOnSpecificDisk) {
```
I suggest you to validate whether `file1.txt` and `file2.txt` files were removed from the `trash` folder and `trashIndex` data structure.

wg1k replied:
It was done. The trashIndex is far from this unit, so the test is in another place.

ralcolea commented on src/chunkserver/chunkserver-common/chunk_trash_manager_impl_unittest.cc:
```diff
@@ -0,0 +1,282 @@
+	std::string timeString = "20231018120350";
+	int errorCode = 0;
+	time_t timeValue =
+	    chunkTrashManagerImpl.getTimeFromString(timeString, errorCode);
```
`getTimeFromString()` function is static, so it can be called directly using `ChunkTrashManagerImpl` class.

wg1k replied:
done

ralcolea commented on src/chunkserver/chunkserver-common/cmr_disk_unittest.cc:
```diff
@@ -0,0 +1,182 @@
+};
+
+std::filesystem::path CmrDiskTest::testDir;
+TEST_F(CmrDiskTest, UnlinkChunkSuccessful) {
```
Some unit tests fail in my local environment.
![image](https://github.com/user-attachments/assets/f08a8bf4-5051-4cbc-ab00-ca953aaaff31)

wg1k replied:
These were expecting the trash to be enabled. Fixed.

ralcolea commented on src/chunkserver/chunkserver-common/CMakeLists.txt:
```diff
@@ -3,6 +3,28 @@ include_directories(${CMAKE_CURRENT_SOURCE_DIR})
-    ${Boost_LIBRARIES} ${ADDITIONAL_LIBS} Boost::system Boost::filesystem)
+        ${Boost_LIBRARIES} ${ADDITIONAL_LIBS})
+
+list(REMOVE_ITEM CHUNKSERVER_PLUGINS_TESTS
```
Why was this change needed?

In my opinion, this change should be avoided if possible. This change isolates `cmr_disk_unittest` into a different binary which is not included in the binary `unittests` generated with all unit tests. Consequently, it requires modifying Github Actions to also run this test or modify the code to include this test in the binary `unittests`.

wg1k replied:
Because of the "-Wl,--wrap=time" below, for avoiding unintended interference with other tests that might rely on the real `time()` function.

wg1k replied:
Let's not affect yet the CI. We'll find a better approach than this patch. I am opening a separated task for fixing testability.

ralcolea commented on src/chunkserver/chunkserver-common/default_disk_manager.cc:
```diff
@@ -28,6 +28,7 @@
 #include "config/cfg.h"
 #include "common/exceptions.h"
 #include "devtools/TracePrinter.h"
+#include "chunk_trash_manager.h"
```
This header is not used and can be safely removed.

wg1k replied:
done

ralcolea commented on src/chunkserver/hddspacemgr.cc:
```diff
@@ -2400,6 +2401,7 @@ void hddFreeResourcesThread() {
 	while (!gTerminate) {
 		gOpenChunks.freeUnused(eventloop_time(), gChunksMapMutex,
 		                       kMaxFreeUnused);
+		ChunkTrashManager::instance().collectGarbage();
```
Static function `collectGarbage()` can be called without `instance()` function.

wg1k replied:
done

ralcolea commented on src/chunkserver/hddspacemgr.cc:
```diff
@@ -2534,6 +2536,7 @@ void hddReload(void) {
 	try {
 		gDiskManager->reloadConfiguration();
 		gDiskManager->reloadDisksFromCfg();
+		ChunkTrashManager::instance().reloadConfig();
```
Static function `reloadConfig()` can be called without `instance()` function.

wg1k replied:
done

ralcolea commented on src/chunkserver/hddspacemgr.cc:
```diff
@@ -2643,15 +2646,30 @@ int hddInit() {
 	try {
 		gDiskManager->reloadConfiguration();
 		gDiskManager->reloadDisksFromCfg();
+		ChunkTrashManager::instance().reloadConfig();
```
Same comment about `reloadConfig()` function.

wg1k replied:
done

ralcolea commented on src/chunkserver/hddspacemgr.cc:
```diff
@@ -2643,15 +2646,30 @@ int hddInit() {
+				                   disk->getPaths().c_str());
+				continue;
+			}
+			ChunkTrashManager::instance().init(disk->metaPath());
```
Same comment about `init()` function.

wg1k replied:
done

ralcolea commented on src/chunkserver/hddspacemgr.cc:
```diff
@@ -2643,15 +2646,30 @@ int hddInit() {
+				continue;
+			}
+			if (disk->metaPath() != disk->dataPath()) {
+				ChunkTrashManager::instance().init(disk->dataPath());
```
Same comment about `init()` function.

wg1k replied:
done

ralcolea commented on src/data/sfschunkserver.cfg.in:
```diff
@@ -184,6 +184,30 @@
+
+## [ADVANCED] The number of files to remove from the trash in a single GC cycle,
+## in case the disk is full and space needs to be recovered. (Default: 100)
+## (Default: 100)
```
Default value appears twice.

wg1k replied:
done

ralcolea reviewed (changes requested):
First of all very good job @ictus4u.
Please, consider my suggestions to improve the final version of this important feature.

uristdwarf commented on doc/sfschunkserver.cfg.5.adoc:
```diff
@@ -153,6 +153,27 @@ the operation is considered failed and is immediately aborted (default: 1000)
+*CHUNK_TRASH_EXPIRATION_SECONDS*:: specifies the timeout in seconds for chunks
+to remain in the trash before being permanently deleted. (Default: 259200)
+
+*CHUNK_TRASH_FREE_SPACE_THRESHOLD_GB*:: sets the available space threshold in
+gigabytes. If the available space on the disk falls below this threshold, the
+system will start deleting older chunks from the trash to free up space. A
+suggested value is about the 15% of the capacity of the smaller configured disk.
+(Default: 0)
```
Minor suggestion: Clarify whether this is per disk or chunkserver.

wg1k replied:
Per drive: `statvfs` is called per drive to check available space. In this first approach, the chunkserver has a single configuration for all of the drives.

uristdwarf commented on src/chunkserver/chunkserver-common/cmr_disk.cc:
```diff
@@ -188,17 +197,63 @@ void CmrDisk::open(IChunk *chunk) {
+		int result = ChunkTrashManager::moveToTrash(
+		    metaFile, metaDiskPath, deletionTime);
+		if (result != SAUNAFS_STATUS_OK) {
+			safs_pretty_errlog(LOG_ERR,
+			                   "Error moving meta file to trash: %s, error: %d",
+			                   metaFile.c_str(), result);
+			return result;
+		}
+
+		// Move data file to trash
+		result = ChunkTrashManager::moveToTrash(
+		    dataFile, dataDiskPath, deletionTime);
+		if (result != SAUNAFS_STATUS_OK) {
+			safs_pretty_errlog(LOG_ERR,
+			                   "Error moving data file to trash: %s, error: %d",
+			                   dataFile.c_str(), result);
+			return result;
+		}
```
I think it should show both the error code and string name of the error. I was going to initially suggest that it be only the string, but our string representations have some missing, so having the exact error code is very helpful also.

wg1k replied:
The log improvements I introduced are going to cover this, but it'll be in another iteration, when these will be refactored to be available out of the `chunk_trash_manager_impl.cc` unit, and public to the whole project.

uristdwarf reviewed (approved):
It would be nice to have some more debugging/tracing logs (e.g when a chunk is moved to trash or deleted)

ralcolea reviewed (approved):
Great job @ictus4u :+1: :fire: :muscle:
